File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/WorldChunkView.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface WorldChunkViewLib: Library {
    fun WorldChunkView_destroy(handle: Pointer)
    fun WorldChunkView_create(cx: Int, cz: Int): Pointer
    fun WorldChunkView_cx(handle: Pointer): Int
    fun WorldChunkView_cz(handle: Pointer): Int
    fun WorldChunkView_to_schematic(handle: Pointer): Pointer
    fun WorldChunkView_set_block(handle: Pointer, x: Int, y: Int, z: Int, blockName: Slice): ResultUnitInt
    fun WorldChunkView_set_biome(handle: Pointer, biomeName: Slice): ResultUnitInt
    fun WorldChunkView_biome_palette_json(handle: Pointer, write: Pointer): ResultUnitInt
}
/** A single decoded chunk (or a from-scratch chunk under construction).
*/
class WorldChunkView internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class WorldChunkViewCleaner(val handle: Pointer, val lib: WorldChunkViewLib) : Runnable {
        override fun run() {
            lib.WorldChunkView_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, WorldChunkView.WorldChunkViewCleaner(handle, WorldChunkView.lib));
    }

    companion object {
        internal val libClass: Class<WorldChunkViewLib> = WorldChunkViewLib::class.java
        internal val lib: WorldChunkViewLib = Native.load("nucleation", libClass)
        @JvmStatic
        
        /** Create an empty chunk view at the given chunk coordinates — the
        *starting point for generating worlds from scratch. Sections are
        *created on demand by `set_block`. Serialized with
        *`status = "minecraft:full"` (Minecraft will not regenerate over it)
        *and the default data version.
        */
        fun create(cx: Int, cz: Int): WorldChunkView {
            
            val returnVal = lib.WorldChunkView_create(cx, cz);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = WorldChunkView(handle, selfEdges, true)
            return returnOpaque
        }
    }
    
    /** The chunk X coordinate (in chunk units).
    */
    fun cx(): Int {
        
        val returnVal = lib.WorldChunkView_cx(handle);
        return (returnVal)
    }
    
    /** The chunk Z coordinate (in chunk units).
    */
    fun cz(): Int {
        
        val returnVal = lib.WorldChunkView_cz(handle);
        return (returnVal)
    }
    
    /** Convert the chunk view to a standalone schematic.
    */
    fun toSchematic(): Schematic {
        
        val returnVal = lib.WorldChunkView_to_schematic(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Schematic(handle, selfEdges, true)
        return returnOpaque
    }
    
    /** Set a block at absolute world coordinates inside this chunk view.
    *`block_name` must be a valid Minecraft block identifier (e.g.
    *`minecraft:stone`). Errors with `InvalidArgument` if (x, z) is outside
    *this chunk's column.
    */
    fun setBlock(x: Int, y: Int, z: Int, blockName: String): Result<Unit> {
        val blockNameSliceMemory = PrimitiveArrayTools.borrowUtf8(blockName)
        
        val returnVal = lib.WorldChunkView_set_block(handle, x, y, z, blockNameSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                return Unit.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            blockNameSliceMemory.close()
        }
    }
    
    /** Overwrite the biome of every currently-present section of the chunk
    *view with `biome_name` (e.g. `minecraft:desert`). Sections are created
    *lazily by `set_block`, so call this AFTER placing blocks.
    */
    fun setBiome(biomeName: String): Result<Unit> {
        val biomeNameSliceMemory = PrimitiveArrayTools.borrowUtf8(biomeName)
        
        val returnVal = lib.WorldChunkView_set_biome(handle, biomeNameSliceMemory.slice);
        try {
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                return Unit.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        } finally {
            biomeNameSliceMemory.close()
        }
    }
    
    /** Deduped union of all sections' biome palette entries, in order of
    *first appearance, written as a JSON array string (`[]` if no section
    *carries biome data).
    */
    fun biomePaletteJson(): Result<String> {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.WorldChunkView_biome_palette_json(handle, write);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            
            val returnString = DW.writeToString(write)
            return returnString.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }

}

File: bindings/php/WorldChunkView.php
<?php
namespace Stencil;

final class WorldChunkView {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned) {
        $this->ptr = $ptr;
        $this->owned = $owned;
    }

    public static function create( $cx,  $cz) {
        $ret = Lib::ffi()->WorldChunkView_create($cx, $cz);
        return new WorldChunkView($ret, true);
    }

    public function cx() {
        $ret = Lib::ffi()->WorldChunkView_cx($this->ptr);
        return $ret;
    }

    public function cz() {
        $ret = Lib::ffi()->WorldChunkView_cz($this->ptr);
        return $ret;
    }

    public function toSchematic() {
        $ret = Lib::ffi()->WorldChunkView_to_schematic($this->ptr);
        return new Schematic($ret, true);
    }

    public static function fromSchematic( $schematic,  $cx,  $cz) {
        $ret = Lib::ffi()->WorldChunkView_from_schematic($schematic->ptr, $cx, $cz);
        return new WorldChunkView($ret, true);
    }

    public function setBlock( $x,  $y,  $z, string $block_name) {
        $__n3 = strlen($block_name);
        $__view3 = Lib::ffi()->new('DiplomatStringView');
        if ($__n3 > 0) {
            $__buf3 = Lib::ffi()->new("uint8_t[" . $__n3 . "]", false);
            \FFI::memcpy($__buf3, $block_name, $__n3);
            $__view3->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf3[0]));
        } else {
            $__view3->data = null;
        }
        $__view3->len = $__n3;
        $result = Lib::ffi()->WorldChunkView_set_block($this->ptr, $x, $y, $z, $__view3);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function setBiome(string $biome_name) {
        $__n0 = strlen($biome_name);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $biome_name, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->WorldChunkView_set_biome($this->ptr, $__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
    }

    public function biomePaletteJson() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->WorldChunkView_biome_palette_json($this->ptr, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->WorldChunkView_destroy($this->ptr);
        }
    }
}
